[KSunil2001] iP

KSunil2001 opened the pull request:
# DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
* easy to learn
+ ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s2.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 

And it is **FREE!**

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

--- 

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Commit:
Greet, Echo, Exit

Commit:
Add, List

Commit:
Mark as Done

Commit:
ToDos, Events, Deadlines

Commit:
Automated Text UI Testing

Commit:
Handle Errors

Commit:
Delete

Commit:
Use Enums

Commit:
Level 7. Save

Commit:
Merge branch 'branch-Level-7'

Commit:
Level 8. Dates and Times

Commit:
Merge branch 'branch-level-8'

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/Storage.java

Commit:
A-MoreOOP: Use More OOP

Commit:
A-Packages: Organize into Packages

Commit:
A-Packages: Fixing previous commit error

Commit:
Merge remote-tracking branch 'origin/add-gradle-support'

Commit:
A-JUnit: Add JUnit Tests

Commit:
A-JavaDoc: Add JavaDoc comments

Commit:
A-CodingStandard: Follow the Coding Standard

Commit:
Level-9: Find

Commit:
Merge branch 'branch-Level-9'

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/Ui.java

# Conflict resolved: Keep both changes

XylusChen commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,44 @@
+    protected String description;
+
+    /** Status of the task */
+    protected boolean isDone;
```
Appreciate you following the naming convention for boolean variables.

XylusChen commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,40 @@
+     */
+    @Override
+    public String toString() {
+        String from = this.from.format(DateTimeFormatter.ofPattern("MMM dd yyyy h:mm a"));
+        String to = this.to.format(DateTimeFormatter.ofPattern("MMM dd yyyy h:mm a"));
```
Perhaps the String conversion of the LocalDateTime objects could be saved as an instance variable? Considering the frequent calls for the toString() method.

XylusChen commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,128 @@
+            while (s.hasNext()) {
+                str = s.nextLine();
+                taskInfo = str.substring(8);
+                if (str.startsWith("T")) {
+                    description = taskInfo;
+                    task = new Todo(description);
+                } else if (str.startsWith("D")) {
+                    int byIdx = taskInfo.indexOf("|");
+                    description = taskInfo.substring(0, byIdx - 1);
+                    LocalDateTime by = LocalDateTime.parse(taskInfo.substring(byIdx + 2), format);
+                    task = new Deadline(description, by);
+                } else {
+                    int fromIdx = taskInfo.indexOf("|");
+                    int toIdx = taskInfo.indexOf("-");
+                    description = taskInfo.substring(0, fromIdx - 1);
+                    LocalDateTime from = LocalDateTime.parse(taskInfo.substring(fromIdx + 2, toIdx - 1), format);
+                    LocalDateTime to = LocalDateTime.parse(taskInfo.substring(toIdx + 2), format);
+                    task = new Event(description, from, to);
```
Would using Switch Case be better here? In view of future feature expansion where more task types could be introduced.

XylusChen commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,93 @@
+                storage.save(tasks.getListOfTasks());
+                break;
+            } catch(StringIndexOutOfBoundsException e) {
+                throw new DukeException(firstWord);
+            }
+        case "deadline":
+            try {
+                int byIdx = command.indexOf("/by");
+                description = command.substring(9, byIdx - 1);
+                LocalDateTime by = LocalDateTime.parse(command.substring(byIdx + 4), format);
+                tasks.saveTask(description, by);
+                storage.save(tasks.getListOfTasks());
+                break;
+            } catch (DateTimeParseException e) {
+                throw new DukeException(e);
+            } catch(StringIndexOutOfBoundsException e) {
+                throw new DukeException(firstWord);
+            }
+        case "event":
+            try {
+                int fromIdx = command.indexOf("/from");
+                int toIdx = command.indexOf("/to");
+                description = command.substring(6, fromIdx - 1);
+                LocalDateTime from = LocalDateTime.parse(command.substring(fromIdx + 6, toIdx - 1), format);
+                LocalDateTime to = LocalDateTime.parse(command.substring(toIdx + 4), format);
+                tasks.saveTask(description, from, to);
+                storage.save(tasks.getListOfTasks());
+                break;
+            } catch (DateTimeParseException e) {
+                throw new DukeException(e);
+            } catch(StringIndexOutOfBoundsException e) {
+                throw new DukeException(firstWord);
+            }
+        case "find":
+            description = command.substring(5);
+            tasks.findTask(description);
+            break;
+        default:
+            throw new DukeException();
+        }
```
Good job on following the expected indentation rule for Switch Case! 💯

XylusChen reviewed (commented):
LGTM. Well-written code that is reasonably readable and comprehendable. Just some minor comments.

Commit:
Fixed some issues on handling errors and OOP implementation

sachindavid96 commented on src/main/java/Duke.java:
```diff
@@ -21,10 +22,52 @@ public static void exit() {
+        String description, by, from, to;
+        String[] input = command.split(" ");
+
+        if (input[0].equalsIgnoreCase("todo")) {
```
I believe a switch statement would look more elegant here

sachindavid96 commented on src/main/java/Duke.java:
```diff
@@ -21,10 +22,52 @@ public static void exit() {
+            end_idx = Arrays.asList(input).indexOf("/from");
+            description = stringConverter(input, start_idx, end_idx);
+
+            start_idx = end_idx + 1;
```
the number 1 could perhaps be defined as a constant named along the lines of 'zero-based indexing adjustment' so that it is very clear to those reading your code

sachindavid96 commented on src/main/java/Duke.java:
```diff
@@ -34,46 +77,62 @@ public static void listTasks() {
         System.out.println("\tHere are the tasks in your list:");
         for (int i = 1; i <= listOfTasks.size(); i++) {
```
this is an opportunity to use an enhanced for-loop

bipbipboopboop commented on src/test/java/duke/task/TodoTest.java:
```diff
@@ -0,0 +1,21 @@
+package duke.task;
+
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
You should move this static import to the top, and use a line to separate it from the normal import.

bipbipboopboop commented on src/test/java/duke/TaskListTest.java:
```diff
@@ -0,0 +1,52 @@
+package duke;
+
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Task;
+import duke.task.Todo;
+import org.junit.jupiter.api.Test;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
```
You should sort the imports here as well.

bipbipboopboop commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,136 @@
+package duke;
+
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Task;
+import duke.task.Todo;
+
+import java.time.LocalDateTime;
+
+import java.util.ArrayList;
```
You should sort the imports here.

bipbipboopboop commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,185 @@
+    /**
+     * Prints the line separator.
+     */
+    public void printLine(){
```
You should fix the brace format here.

bipbipboopboop reviewed (commented):
Overall, your code looks good. There are some format errors, you can try to setup your IDE to auto format the code for you.

Commit:
Implement GUI for Duke

Commit:
Merge branch 'branch-Level-10'

Commit:
Add assertions to Duke

There are no assertions to document important assumptions that should
hold at various points in the code.

Need to document important assumptions that should hold at
various points in the code so that the runtime can verify them.

Add assertions to Duke.java, Storage.java and TaskList.java.

This is done because an assertion failure indicates a possible bug and
the sooner the execution stops, the safer it is.

Commit:
Refactor the code to improve code quality

The code has too long methods, magic numbers, complicated expressions
and not enough use of SLAP.

The code needs to be refactored to improve its understandability.

Use SLAP to cut down long methods, simplify complicated expressions in
steps and use named constants for magic numbers.

This improves the code quality and thus its understandability.

KSunil2001 committed:
Merge pull request #3 from KSunil2001/branch-A-Assertions

Add assertions to Duke

Commit:
Merge branch 'master' of https://github.com/KSunil2001/ip

* 'master' of https://github.com/KSunil2001/ip:
  Add assertions to Duke

Commit:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/TaskList.java

KSunil2001 committed:
Merge pull request #4 from KSunil2001/branch-A-CodeQuality

Refactor the code to improve code quality

Commit:
Merge branch 'master' of https://github.com/KSunil2001/ip

* 'master' of https://github.com/KSunil2001/ip:
  Refactor the code to improve code quality

Commit:
Update main class name in build.gradle

Commit:
Add help command to Duke

Commit:
Merge branch 'branch-BCD-Extension'

KSunil2001 committed:
Update README.md

Commit:
Fix minor issues

Commit:
Merge branch 'master' of https://github.com/KSunil2001/ip

* 'master' of https://github.com/KSunil2001/ip:
  Update README.md

Commit:
Fix some more minor issues

Commit:
Fix file loading error
